packages/react-reconciler/src/ReactFiberBeginWork.js

```diff
@@ -605,6 +610,222 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
+              nextFiber = null;
+              break;
+            }
+            let sibling = nextFiber.sibling;
+            if (sibling !== null) {
+              nextFiber = sibling;
+              break;
+            }
+            // No more siblings. Traverse up.
+            nextFiber = nextFiber.return;
+          }
+        }
+        fiber = nextFiber;
+      }
+    }
+  }
+
+  function updateProviderComponent(
+    current,
+    workInProgress,
+    renderExpirationTime,
+  ) {
+    if (enableNewContextAPI) {
+      const providerType: ReactProviderType<any> = workInProgress.type;
+      const context: ReactContext<any> = providerType.context;
+
+      const newProps = workInProgress.pendingProps;
+      const oldProps = workInProgress.memoizedProps;
+
+      pushProvider(workInProgress);
+
+      if (hasLegacyContextChanged()) {
+        // Normally we can bail out on props equality but if context has changed
+        // we don't do the bailout and we have to reuse existing props instead.
+      } else if (oldProps === newProps) {
+        return bailoutOnAlreadyFinishedWork(current, workInProgress);
+      }
+      workInProgress.memoizedProps = newProps;
+
+      const newValue = newProps.value;
+
+      let changedBits: number;
+      if (oldProps === null) {
+        // Initial render
+        changedBits = MAX_SIGNED_32_BIT_INT;
+      } else {
+        const oldValue = oldProps.value;
+        // Use Object.is to compare the new context value to the old value.
+        if (!is(newValue, oldValue)) {
+          changedBits =
+            context.calculateChangedBits !== null
+              ? context.calculateChangedBits(oldValue, newValue)
+              : MAX_SIGNED_32_BIT_INT;
+          if (changedBits !== 0) {
+            propagateContextChange(
+              workInProgress,
+              context,
+              changedBits,
+              renderExpirationTime,
+            );
+          }
+        } else {
+          // No change.
+          changedBits = 0;
+        }
+      }
+      workInProgress.stateNode = changedBits;
+
+      if (oldProps !== null && oldProps.children === newProps.children) {
+        return bailoutOnAlreadyFinishedWork(current, workInProgress);
+      }
+      const newChildren = newProps.children;
+      reconcileChildren(current, workInProgress, newChildren);
+      return workInProgress.child;
+    } else {
+      return null;
+    }
+  }
+
+  function updateConsumerComponent(
+    current,
+    workInProgress,
+    renderExpirationTime,
+  ) {
+    if (enableNewContextAPI) {
+      const context: ReactContext<any> = workInProgress.type;
+
+      const newProps = workInProgress.pendingProps;
+      const oldProps = workInProgress.memoizedProps;
+
+      // Get the nearest ancestor provider.
+      const providerFiber: Fiber | null = context.currentProvider;
+
+      let newValue;
+      let changedBits;
+      if (providerFiber === null) {
```

Comment:
This is a hot path and we want these to be able to be sprinkled all over the app so they must be cheap. We also expect the one to many ratio to be big. So whatever work we can do in the provider is worth it.

This whole things could just be:

```js
let newValue = context.currentValue;
let changedBits = context.changedBits;
```

All we need is two new fields on the context type and to set them in the provider (and to reset them when we clear).